File: raspberry/SmartHome/thirdpartydevices/src/main/java/raspberry/smarthome/thirdpartydevices/xiaomi/gateway/device/WeatherSensor.java
package raspberry.smarthome.thirdpartydevices.xiaomi.gateway.device;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Optional;

import raspberry.smarthome.thirdpartydevices.R;

public class WeatherSensor extends THSensor {

    private OnPressureChangeListener listener;

    private Float pressure = null;

    public WeatherSensor(String sid) {
        super(sid, WEATHER_SENSOR_TYPE);
    }

    public void setListener(OnPressureChangeListener listener) {
        this.listener = listener;
    }

    @Override
    public void parseData(String cmd) {
        super.parseData(cmd);

        try {
            JSONObject o = new JSONObject(cmd);

            if(!o.isNull(STATUS_PRESSURE)) {
                float currPress = Float.parseFloat(o.getString(STATUS_PRESSURE)) / 100;

                if(pressure != null && Math.abs(currPress - pressure) > 0.01) {
                    Optional.ofNullable(listener).ifPresent(onPressureChangeListener -> onPressureChangeListener.onPressureChanged(currPress));
                }

                pressure = currPress;
            }

        } catch (JSONException | NumberFormatException e) {
            e.printStackTrace();
        }
    }

    @Override
    public String toString() {
        return super.toString() + "pressure: " + pressure + "kPA";
    }

    @Override
    public Bitmap getDevicePicture(Resources resources) {
        return BitmapFactory.decodeResource(resources, R.drawable.xiaomi_thp_sensor);
    }

    public interface OnPressureChangeListener {

        void onPressureChanged(float p);
    }
}

File: raspberry/SmartHome/thirdpartydevices/src/main/java/smarthome/raspberry/thirdpartydevices/xiaomi/gateway/device/WeatherSensor.kt
package smarthome.raspberry.thirdpartydevices.xiaomi.gateway.device

import org.json.JSONException
import org.json.JSONObject
import smarthome.library.common.constants.DeviceTypes.WEATHER_SENSOR_TYPE
import smarthome.library.common.constants.GATEWAY_HUMIDITY_CONTROLLER
import smarthome.library.common.constants.GATEWAY_PRESSURE_CONTROLLER
import smarthome.library.common.constants.GATEWAY_TEMPERATURE_CONTROLLER
import smarthome.raspberry.thirdpartydevices.xiaomi.gateway.constants.STATUS_PRESSURE
import smarthome.raspberry.thirdpartydevices.xiaomi.gateway.controller.PressureController

class WeatherSensor(sid: String, gatewaySid: String)
    : THSensor(sid, WEATHER_SENSOR_TYPE, gatewaySid) {

    init {
        addControllers(PressureController(this))
    }

    override fun parseData(json: String) {
        super.parseData(json)
        try {
            val o = JSONObject(json)

            if (!o.isNull(STATUS_PRESSURE))
                getControllerByType(GATEWAY_PRESSURE_CONTROLLER).state = (o.getString(STATUS_PRESSURE).toFloat() / 1000).toString() + "kPa"

            super.parseData(json)
        } catch (e: JSONException) {
            reportDataParseError(e)
        }
    }

    override fun toString(): String {
        return super.toString() + "\ntemperature: " + getControllerByType(GATEWAY_TEMPERATURE_CONTROLLER).state +
                ", humidity: " + getControllerByType(GATEWAY_HUMIDITY_CONTROLLER).state +
                ", pressure: " + getControllerByType(GATEWAY_PRESSURE_CONTROLLER).state
    }
}